### ISSUE TITLE
When opening a bigger than memory CSV, using chunksize and converting to HDF5, the memory is still fully used until the machine hangs and the kernel dies.

### ISSUE THREAD

**Issue by nf78** [vaexio/vaex#827]
Following the recommendation here:
https://vaex.readthedocs.io/en/latest/faq.html

I have been able to verify that the small files (from each chunk) are being generated. 

However, although the data is being chunked and the small files are created, the memory usage increases until all memory is used, and the machine hangs until the kernel dies.

When reading the dataset using the chunksize argument, into pandas.read_csv(), the memory is kept as low as the chunksize. Perhaps the memory leak is happening in the process of conversion to hdf5?

To reproduce the issue, just try to read a bigger than memory csv file with this code (as recommended on the faq link):

df = vaex.from_csv('./my_data/my_big_file.csv', convert=True, chunk_size=5_000_000)

**Comment by nf78** [comment]
Also, I want to add that, once all memory is fully used, no more small hdf5 files are created during conversion

**Comment by byaminov** [comment]
@nf78 thanks for raising the issue!

I've tried converting 2.2Gb CSV file on a docker container limited to 1Gb memory, and it worked. I had to use smaller `chunk_size=100_000`.

However I could reproduce the issue when the CSV file contains non-number columns: the chunks are created successfully, but when trying to create one HDF5 file from the chunks the process crashes in that docker machine.

Does your CSV file also contain string columns? If not, can you try reducing the `chunk_size`? If yes, then it's probably the same issue that I could reproduce. We'll be looking into it further.

**Comment by byaminov** [comment]
For the record, the command I used to start the Docker image:
```
docker run -it \
    --memory="1g" \
    --memory-swap="1g" \
    -v /my/datasets/test_yellow_tripdata:/data/ \
    python:3.7 /bin/bash
```

**Comment by nf78** [comment]
Hi @byaminov ! Thanks for looking into this.

I should also update that my working environment is on an EC2 with 32GB and the dataset is a CSV with 54GB, running Ubuntu 18.04.2 LTS and python 3.6.9. The version of vaex is the latest available through pip install.

With regards to your questions:
a) Does your CSV file also contain string columns? 
     > Yes, it is mostly text.

b) If yes, then it's probably the same issue that I could reproduce. We'll be looking into it further.
      > Thank you

**Comment by nf78** [comment]
By the way, I just noticed on your docker container command, you specify 1GB of RAM memory and 1GB of SWAP, adding those two together, the total memory available in your system is close to the 2.2GB size of your CSV file (as a rule of thumb, CSV filesize is highly correlated to RAM allocation) 

In my case there is no SWAP setup on my EC2, just 32GB of RAM.

**Comment by nf78** [comment]
I would like to work on this issue, if that's ok?

**Comment by JovanVeljanoski** [comment]
Yes, please feel free!

**Comment by nf78** [comment]
Thanks!

I've managed to fix the issue and raised the PR#887, however there are a couple of unit-tests failing, that are not part of my fix:

=========================== short test summary info ============================
FAILED tests/jupyter/model_test.py::test_model_status[df_trimmed-tornado_client]
FAILED tests/jupyter/model_test.py::test_model_status[df_remote-tornado_client]
= 2 failed, 1858 passed, 26 skipped, 8 xfailed, 4 xpassed, 148 warnings in 440.82s (0:07:20) =
The command "py.test tests/ ${PYTEST_EXTRA_ARGS}" exited with 1.

**Comment by nf78** [comment]
Hi, this failing tests surely shouldn't be in the repo, right? I wonder how this failing tests managed to get merged?

**Comment by maartenbreddels** [comment]
Hi @nf78,

just letting you know I'm busy, hope to take a look at your work ASAP. That test is a known issue, it fails only on CI, not my local machine. The failure is most likely not an issue (it has to do with asyncio timing). Don't worry about that, it's pushed down a bit on my TODO list.  https://github.com/vaexio/vaex/pull/826 introduced the failure AFAIK.

cheers,

Maarten

**Comment by nf78** [comment]
Thanks for your reply @maartenbreddels, and taking the time to explain the situation, considering you're busy, really that's appreciated.

I hope to get the chance to contribute even further in the future to this great project.

Looking forward to hearing from you.

**Comment by byaminov** [comment]
> By the way, I just noticed on your docker container command, you specify 1GB of RAM memory and 1GB of SWAP, adding those two together, the total memory available in your system is close to the 2.2GB size of your CSV file

@nf78 yep the Docker command looked misleading to me too, but according to [Docker docs](https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details) it is the way to allocate 1GB of total RAM with no swap: 

> * If --memory-swap is set to a positive integer, then both --memory and --memory-swap must be set. --memory-swap represents the total amount of memory and swap that can be used, and --memory controls the amount used by non-swap memory. So if --memory="300m" and --memory-swap="1g", the container can use 300m of memory and 700m (1g - 300m) swap.
> 
> PREVENT A CONTAINER FROM USING SWAP
> If --memory and --memory-swap are set to the same value, this prevents containers from using any swap. This is because --memory-swap is the amount of combined memory and swap that can be used, while --memory is only the amount of physical memory that can be used.

**Comment by nf78** [comment]
Ahh, makes sense now, thanks @byaminov for clarifying!

**Comment by lrq3000** [comment]
Sorry but is there any news on this? If I understand correctly, reducing the chunk_size should allow to process these files without a memory overflow, but if the PR could be merged then it would fix this rather limiting issue for those doing big data on files containing strings :-)

**Comment by JovanVeljanoski** [comment]
Can you try the newest alpha?

**Comment by nf78** [comment]
> Can you try the newest alpha?

Hi @JovanVeljanoski, I'm not sure which alpha tag you refer to, there is the core and the hdf5 versions, both are alpha.

By the way, I've got a PR since June, and I've got no news since then. 

Kind Regards

**Comment by maartenbreddels** [comment]
Hi,

this issue is not forgotten. I've spent time looking into this, but memory_profiler seems to give very confusing results. The latest vaex-core alpha did solve a memory issue, so I'm wondering if it is fixed. I do see high memory usage, but I wonder if it's real, is memory_profiler doesn't count the memory-mapped memory to it's total.

**Comment by maartenbreddels** [comment]
FYI: I can basically now make memory usage very small by commenting out this line:
https://github.com/vaexio/vaex/blob/098658c9783634f54625c25f3780641f06f68a0e/packages/vaex-core/vaex/export.py#L217

Which is simply writing to a memory mapped, which means to me that whatever memory_profiler is reporting is a bit misleading.

Looking at https://github.com/pythonprofilers/memory_profiler/blob/bd6da910f791cef725640fa207fb4ee433c93586/memory_profiler.py#L132 it seems it is reporting RSS memory. Looking at https://unix.stackexchange.com/a/535071 indeed it looks like this include the pages/mmaped memory being written to. Note that this memory is not really 'used', since it can be discarded at any moment by the OS. So even though memory usage seems high, it should not run of out memory

TLDR: I think this is solved, but I'd like to see this confirmed.

**Comment by nf78** [comment]
> Hi,
> 
> this issue is not forgotten. I've spent time looking into this, but memory_profiler seems to give very confusing results. The latest vaex-core alpha did solve a memory issue, so I'm wondering if it is fixed. I do see high memory usage, but I wonder if it's real, is memory_profiler doesn't count the memory-mapped memory to it's total.

Hi @maartenbreddels ,

Thanks for looking into this. With regards to the confusing results of memory_profiler, it sometimes happen if the threshold for the garbage collection is set higher than the number of memory allocations minus the number of deallocations. Because garbage collection only starts once this difference is higher than the threshold.

This can be reduced with the gc.set_threshold() method.

So to answer your question, if memory_profiler sees high memory consumption, it is real. And you can see when the garbage collection is done (manually done in this case at critical points), the memory drops.

However, I am not sure if you need me to go and test the vaex-core alpha, because you are still seeing the high memory usage?

**Comment by maartenbreddels** [comment]
It's not real memory usage, it's mmaped data being used. If I enable this code:
https://github.com/vaexio/vaex/blob/098658c9783634f54625c25f3780641f06f68a0e/packages/vaex-core/vaex/dataset_mmap.py#L92

It will not use any memory after the export (since the mapped file is properly closde), but for some reason this makes it crash in CI (I'll try to enable it again, and see if maybe that is not the case anymore).

I think it really is fixed, and what you've seen before, was another arrow related issue. So I think it was a real issue, that was being conflated with the memory mapping showing memory usage as well.

**Comment by nf78** [comment]
> It will not use any memory after the export (since the mapped file is properly closde), but for some reason this makes it crash in CI (I'll try to enable it again, and see if maybe that is not the case anymore).
> 
> I think it really is fixed, and what you've seen before, was another arrow related issue. So I think it was a real issue, that was being conflated with the memory mapping showing memory usage as well.

Ok, thanks for clarifying, I now get a better picture of what you are saying. So do you want me to go and test it with the latest vaex-core alpha?

**Comment by maartenbreddels** [comment]
That would be great if you could do that!

**Comment by nf78** [comment]
> That would be great if you could do that!

cool, I'll look into it asap.

**Comment by maartenbreddels** [comment]
Ok, I was curious if I could check this as well using docker, so here are my steps:
```
# issue with docker that cannot disable swap
$ sudo swapoff -a 
# run docker with limited amount of memory, mount my dev conda environment, the data dir, and the vaex dev dir
$ docker run -it -v /data/:/data -v /home/maartenbreddels/miniconda:/root/miniconda -v /home/maartenbreddels/:/home/maartenbreddels/ --memory=312m --memory-swap=0 ubuntu
# inside the docker image:
export PATH=/root/miniconda/envs/dev/bin:$PATH
export PYTHONPATH=/home/maartenbreddels/github/vaexio/vaex/packages/vaex-core
python /home/maartenbreddels/github/vaexio/vaex/memleak.py 
```

Where memleak.py contains:
```python
import vaex
df = vaex.from_csv('/data/tmp/memleak.csv', convert=True, chunk_size=100_000)
```

passing `--memory=112m` will make it fail, but `312m` is enough to make it run, even when memory_profiler says `1504.1 MB` is used.

So I'm pretty sure it works with the data I have but still would love to get confirmation from you.

**Comment by nf78** [comment]
> conda environment, the data dir, and the vaex dev dir

Thanks for sharing this information, I will look into it

**Comment by nf78** [comment]
> Ok, I was curious if I could check this as well using docker, so here are my steps:
> 
> ```
> # issue with docker that cannot disable swap
> $ sudo swapoff -a 
> # run docker with limited amount of memory, mount my dev conda environment, the data dir, and the vaex dev dir
> $ docker run -it -v /data/:/data -v /home/maartenbreddels/miniconda:/root/miniconda -v /home/maartenbreddels/:/home/maartenbreddels/ --memory=312m --memory-swap=0 ubuntu
> # inside the docker image:
> export PATH=/root/miniconda/envs/dev/bin:$PATH
> export PYTHONPATH=/home/maartenbreddels/github/vaexio/vaex/packages/vaex-core
> python /home/maartenbreddels/github/vaexio/vaex/memleak.py 
> ```
> 
> Where memleak.py contains:
> 
> ```python
> import vaex
> df = vaex.from_csv('/data/tmp/memleak.csv', convert=True, chunk_size=100_000)
> ```
> 
> passing `--memory=112m` will make it fail, but `312m` is enough to make it run, even when memory_profiler says `1504.1 MB` is used.
> 
> So I'm pretty sure it works with the data I have but still would love to get confirmation from you.

Hi @maartenbreddels can you please confirm if you use any particular docker image that I should pull? or any Dockerfile? I couldn't find anything on the vaex docs, sorry. And I noticed you have a container named "ubuntu" but not sure if this is 18.04 or 20.04 It also looks like you use miniconda instead of python, also not sure which python version? Thanks in advance for clarifying

**Comment by maartenbreddels** [comment]
you should be able to run the command as it, it will take the latest ubuntu. There is nothing in the docs about this, it is simply an experiment I tried. You can simply mount your local python environment in the docker container at the same location to avoid having to install it in the container.

**Comment by nf78** [comment]
Hi @maartenbreddels,

I've tried to run the test with the latest alpha, but I'm running on this issue:

`
RuntimeError: Oops, failed to install vaex-core
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/blackteam/dev-vaex/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/blackteam/vaex/setup.py'"'"'; __file__='"'"'/home/blackteam/vaex/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
`

Here is my procedure, please let me know if I'm missing anything:

mkdir dev-vaex
virtualenv dev-vaex
source dev-vaex/bin/activate
git clone https://github.com/vaexio/vaex.git
cd vaex
git checkout tags/core-v4.0.0-alpha.10
pip install -e .

**Comment by maartenbreddels** [comment]
Hi,

That was a bit of a short error msg :), but I guess you missed the recursive flag `git clone --recursive https://github.com/vaexio/vaex ` ( see https://vaex.io/docs/installing.html#for-developers )

**Comment by nf78** [comment]
@maartenbreddels, 

hehe, yeah I figured you wouldn't need the extensive version of the error message to tell what the problem could be :) 

Thanks for pointing me to the installing docs link, I was just doing the standard stuff so that helped a lot. 

Also, I would like to add, I figured from those docs I was also missing the pcre package, so had to do:

`apt-get install libpcre3 libpcre3-dev`

After that it installed like a charm :) 

Anyway, I should be now able to test this soon. Will get back to you

**Comment by nf78** [comment]
Hi @maartenbreddels ,

Sorry for taking a bit to get back to you, considering Xmas holidays and I've also been struggling to get vaex to work in a virtualenv with vaex, using containers.

In a nutshell, I would like to know all the steps you follow to set up your environment in local or container.

What I have tried so far is:

a) Create the virtualenv in a dev-vaex folder on local machine, and then install the alpha version of vaex (also in local). This install correctly but this is not recognized inside the container, when I use the commands you have provided. So I must be missing something here.

b) For trial sake, I have also attempted to install the virtualenv and the vaex alpha inside the container, but this just errors, when trying to install vaex in editing mode.

I've followed all steps in documentation, but I'm still blocked and not sure what I'm missing here.

## PROCEDURE: Install in LOCAL
cd /media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76
mkdir dev-vaex
mkdir vaex

## preparing environment:
apt-get install -y libpcre3 libpcre3-dev
python3 -m pip install --upgrade pip
pip3 install virtualenv
python3 -m virtualenv dev-vaex

## now installing vaex in editing mode
git clone --recursive https://github.com/vaexio/vaex
cd vaex
git checkout tags/core-v4.0.0-alpha.11
source /media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/dev-vaex/bin/activate
pip3 install -e .

### run docker with limited amount of memory, mount the data dir, my dev python environment and the vaex dev dir
sudo docker run -it -v /media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/data:/data -v /media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/dev-vaex:/home/ubuntu/dev-vaex -v /media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/vaex:/home/ubuntu/vaex --memory=312m --memory-swap=0 ubuntu:20.04

# inside the docker image:
## many packages are not installed by default on ubuntu docker image:
apt-get update
apt-get install -y python3 python3-pip

## adding local folders to path
export PATH=/media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/dev-vaex/bin:$PATH
export PYTHONPATH=/media/ubuntu/2ac10185-740d-4a6b-8237-9f40d0010a76/vaex/packages/vaex-core

At this point, when I activate the virtual environment inside the container, it seems to work with "import vaex" but no other api call after that. 

Also pip3 show vaex, shows "WARNING: Package(s) not found: vaex"

**Comment by nf78** [comment]
Hi @maartenbreddels !

I was wondering if you had the chance to read my previous comments?

Thanks.

**Comment by nf78** [comment]
@maartenbreddels @byaminov @JovanVeljanoski 

I was wondering if anyone had the chance to read my previous comments?

Thanks.

**Comment by maartenbreddels** [comment]
I'm not sure what the issue is, but it's unlikely this is vaex related. Did you try this with a really simple package?

**Comment by maartenbreddels** [comment]
I am assuming for now this issue is solved, and we can also close https://github.com/vaexio/vaex/pull/887. If not, please reopen. Thanks for digging into this, I hope you can solve your installation issues!

**Comment by nf78** [comment]
Hi @maartenbreddels , sorry I didn't have the chance to look into this any further recently. Once I get the chance to test it, I will let you know.

**Comment by werbliben** [comment]
@maartenbreddels @nf78 just want to confirm that the issue is not resolved as of version 4.3.0. A larger-than-memory CSV with string columns read with `df = vaex.from_csv('./my_data/my_big_file.csv', convert=True, chunk_size=5_000_000)` is indeed cut into hdf5 chunks, but the kernel crashes after the last chunk is written to disk.

Is it possible to reopen the issue?

**Comment by JovanVeljanoski** [comment]
hi, 

can you please check if all the chunks are written correctly, and if the final large single file (that should contain all of the data in one file) is also created?

Also can you give us more information about the data you are using and the system itself? For instance the number of columns, the types, typical/average length of the string columns etc.. We would need to reproduce it. 

I used vaex to convert a 90 GB csv file (mixed numerical and string columns, tho more numerical), and it worked fine.

**Comment by werbliben** [comment]
@JovanVeljanoski 

thanks for reacting so quickly!

The final, merged, file was indeed created -- I somehow glanced over it at first, my bad. When I import it back into the notebook, the number of rows seems to be correct, but all but the first column are empty.

There are 11 columns in total, nine string columns (incl. the first one) and two numerical. The CSV is 65 GB and ~270 mil. rows long. Strings in one of the columns are Reddit comments and can get pretty long (longest one is 1285 words), but the average is quite low at only 5.6 words.

The machine is a VM running on a node with two 32-core AMD EPYC 2's and 900 GB of memory.

**Comment by maartenbreddels** [comment]
Could you run it from a script, and tell what the final error msg is? Bus
error or seg fault? Or killed(might indicate out of memory)

(from mobile phone)

**Comment by werbliben** [comment]
@maartenbreddels

Ran a script. There... is no error message, but something seems off. After the line which opens the DF (`df = vaex.from_csv('./my_data/my_big_file.csv', convert=True, chunk_size=5_000_000)`), I added the following two:

```
print(df.head())
input("Finished writing, press Enter to exit.")
```

Neither of the above lines seems to get executed, the script just silently exits after concatenating all the chunks into the final HDF5 file. On top of that, no chunks are removed after the merge, though the [documentation](https://vaex.readthedocs.io/en/latest/example_io.html) seems to imply that they should:

> All temporary files are then concatenated into a single HDF5 file, and the temporary files deleted.

So it seems that the script silently exits at the step at which the chunks should be removed.

**Comment by JovanVeljanoski** [comment]
Was the file correctly created however? Earlier you said that some data was missing? Are the individual chunks correct (you can check the first and the last for example).

**Comment by maartenbreddels** [comment]
Did you check with dmesg if the OOM killer killed it? It's weird that it
silently quits, nothing in vaex should do that, really strange behaviour!
Did you try a different version, maybe make a new conda environment?

(from mobile phone)
